### PR TITLE
Change map when copying Roaring64MapSetBitForwardIterator

### PR DESCRIPTION
**Pull request by tudor**
See https://github.com/RoaringBitmap/CRoaring/issues/589.

Fixed by turning the map reference into a pointer, in which case the default copy assignment operator will do the right thing. (`std::reference_wrapper` would also have worked, but I find that way too pedantic.)

Now that we did this, we don't need `map_begin` and `map_end` any more -- just call `p->cbegin()` and `p->cend()` as needed. (map iterators are very cheap to construct)

Now that we did this, we don't need the difference between `Roaring64MapSetBitForwardIterator` and `Roaring64MapSetBitBiDirectionalIterator` -- I made`Roaring64MapSetBitBiDirectionalIterator` the only iterator and added typedefs as needed for backwards compatibility. 

Added a test, all tests pass.

### CODE REVIEW

File: tests/cpp_unit.cpp
Comment (SLieve) [inline tests/cpp_unit.cpp:2149]: Could you add a brief comment about the purpose of the test?

File: cpp/roaring64map.hh
Comment (SLieve) [inline cpp/roaring64map.hh:1836]: Please move to the top of the class.

File: cpp/roaring64map.hh
Comment (SLieve) [inline cpp/roaring64map.hh:1713]: Why `final`?

File: cpp/roaring64map.hh
Comment (SLieve) [inline cpp/roaring64map.hh:35]: I think it would be useful to leave a comment about the purpose of this typedef, given that it's no longer used elsewhere in this file.

File: tests/cpp_unit.cpp
Comment (tudor) [reply tests/cpp_unit.cpp:2151]: Done

File: cpp/roaring64map.hh
Comment (tudor) [reply cpp/roaring64map.hh]: Done

File: cpp/roaring64map.hh
Comment (tudor) [reply cpp/roaring64map.hh]: Removed

File: cpp/roaring64map.hh
Comment (tudor) [reply cpp/roaring64map.hh:38]: Done

**Comment by tudor** [comment]
@SLieve PTAL, thank you!

**Review by SLieve — APPROVED** [review]

**Comment by tudor** [comment]
@SLieve @lemire I can't merge this myself as I don't have write access but it's been approved. Thank you!

**Pull request merged**
